Add duplicateLetter exercise + implementation

vaibhavarora14 opened the pull request:
**Program Strategy -**
To find out most duplicate letter `count` from a `string`, program does following things:
- With `Set`, unique letters copy is pulled out from `string`.
- For every `unique letter`, a filter runs on 'string' to find out that `letter` length.
    - If letter length is greater than, we found previously, it's `stored`.
    - With this, `stored` count in end is count of maximum duplicate `letter count` from `string`.

Commit:
WIP: Add duplicateLetter exercise + implementation

varora1406 commented:
Performance of program is not good according to me.
Working on it's performance.

Commit:
Add exceptions && remove ToDo
- Exceptions are added in case contract of duplicateLetters input is not fullfiled.
- Removing ToDo, found program to be enough performant in analysis

varora1406 commented:
Program is ready for review

ahhda commented on js-exercises/duplicate-letters/duplicateLetters.js:
```diff
@@ -0,0 +1,49 @@
+
+    if (letterCount > mostDuplicateLetter.count) {
+      mostDuplicateLetter.count = letterCount;
+      mostDuplicateLetter.letter = letter;
```
Do we really need to store the letter as well? Just trying to understand your thought process. Were you trying to implement thinking that in the future you might need it?

varora1406 replied:
This program just needed count, but I went with real time expectation to keep letter as well in program.
Will remove from here for the time, and keep it fully according to specs

varora1406 replied:
it's removed with https://github.com/pesto-students/deep-dive-varora1406/pull/1/commits/f47a8012116b60b63ef809392c344e14ec1e118f

Commit:
Merge branch 'master' of https://github.com/pesto-students/deep-dive-varora1406 into duplicate-letters

Commit:
shorten LOC of emptyString()

Commit:
Remove letter property of mostDuplicateLetter

Commit:
Remove letter property from mostDuplicateLetter

Commit:
Bringing filter to single line for readability

varora1406 commented on js-exercises/duplicate-letters/duplicateLetters.js:
```diff
@@ -0,0 +1,49 @@
+    return true;
+  }
+
+  return false;
```
improved with https://github.com/pesto-students/deep-dive-varora1406/pull/1/commits/47178678036add39e53c5f851b743f4bd3f29b3d

varora1406 commented on js-exercises/duplicate-letters/duplicateLetters.js:
```diff
@@ -0,0 +1,52 @@
+
+const letterRegex = /^[A-Za-z]+$/;
+
+const isLetter = (letter) => letterRegex.test(letter);
```
Actually by `letter`, I meant an `alphabet`.
So letterRegex should neither be true for `abc def` nor even it should be true for `abc` OR `def`.

With this, I realize isLetter currently not handling length of letters. will update this in solution soon

varora1406 replied:
it's available with 5b203a656feb2bfee9b9d66bfe59c9d8fdcd9937

Commit:
Update letter regex with 1 length only

varora1406 commented on js-exercises/duplicate-letters/duplicateLetters.js:
```diff
@@ -0,0 +1,52 @@
+    return false;
+  }
+
+  return mostDuplicateLetter.count;
```
just wanted to make code easily readable.
If we can ignore LOC, and you can share the way you would like to see the code for your eyes.

 I will be happy to receive that feedback. 🙂

1akshat replied:
yes definitely understand that.

varora1406 commented on js-exercises/duplicate-letters/duplicateLetters.js:
```diff
@@ -0,0 +1,52 @@
+    if (!isLetter(letter)) {
+      mostDuplicateLetter.count = 0;
+      mostDuplicateLetter.letter = null;
+      break;
```
Yes, with latest code.
The `if` block now throws error actually. so both of our code block won't fit there now 😃.
Have a look at it 😉

1akshat replied:
sure.

varora1406 commented:
Guys, I have updated subject with code strategy used in this program
